### PR TITLE
Block Hooks: Allow traversal callbacks to modify parent block

### PR DESCRIPTION
**Pull request by ockham**
_tl;dr:_ In 6.5, we'll need arguments to two functions introduced during the 6.4 cycle to become references. Since that can be deemed a back-compat breaking change, I'm suggesting to make that change already.

----

The callbacks returned by `make_before_block_visitor` and `make_after_block_visitor`, respectively (which are passed as arguments to `traverse_and_serialize_block(s)`) currently accept three arguments, all of which are block arrays (i.e. with properties `blockName`, `attrs`, etc.):
- A _reference_ to the block they're currently visiting, `&$block`;
- the block's `$parent_block`; and
- the `$prev`ious block (for `make_before_block_visitor`), or the `$next` block (for `make_after_block_visitor`), respectively.

Those arguments are passed to the "block visitor" callbacks by `traverse_and_serialize_block(s)` during traversal. The block that the callback is currently visiting is passed _by reference_ to allow modifying it, which is e.g. used to inject the `theme` attribute into Template Part blocks.

One major limitation with Block Hooks is that they currently only work with templates, parts, and patterns that _don't have any user modifications_ (i.e. that come straight from the corresponding theme files, rather than the DB). For WP 6.5, we're planning to change that to also make Block Hooks work for templates, parts, and patterns that _do_ have user modifications: [`#59646`](https://core.trac.wordpress.org/ticket/59646).

The way we'll make this work is by storing an attribute on the "anchor" block. This is currently being developed in https://github.com/WordPress/wordpress-develop/pull/5523.

While working on that PR, I noticed that that means that we'll also need to allow the aforementioned callbacks to modify not only the currently visited `$block`, but also the `$parent_block` -- i.e. to pass the latter argument by reference, too. This is consistent with the requirement of adding an attribute to an anchor block, as it's not only the currently visited block that can serve as an anchor block (in the case of `before` or `after` sibling insertion), but also its parent (for `first_child` and `last_child` insertion).

Per [discussion](https://wordpress.slack.com/archives/C02RQBWTW/p1698765741579409) with @hellofromtonya, changing the `$parent_block` argument to become a reference can be considered a backwards-compatibility breaking change. I'm thus suggesting to make that change already as part of 6.4, which is the cycle where the relevant functions were first introduced. This should have no impact on existing code, since nothing currently relies on `$parent_block` remaining unmodified by the respective callback, nor is anything currently modifying that argument.

Trac ticket: https://core.trac.wordpress.org/ticket/59776

---
**This Pull Request is for code review only. Please keep all other discussion in the Trac ticket. Do not merge this Pull Request. See [GitHub Pull Requests for Code Review](https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/) in the Core Handbook for more details.**

### CODE REVIEW

File: src/wp-includes/blocks.php [hunk 1]
```diff
@@ -785,7 +785,7 @@ function make_before_block_visitor( $hooked_blocks, $context ) {
 	 * @param array $prev         The previous sibling block of the given block.
 	 * @return string The serialized markup for the given block, with the markup for any hooked blocks prepended to it.
 	 */
-	return function ( &$block, $parent_block = null, $prev = null ) use ( $hooked_blocks, $context ) {
+	return function ( &$block, &$parent_block = null, $prev = null ) use ( $hooked_blocks, $context ) {
```
Comment (hellofromtonya) [inline src/wp-includes/blocks.php:788]: @ockham should also document both in the `@param` as passed by reference.

Core has 2 different ways of documenting this at the end of the `@param` string:

* `(passed by reference).`
* `. Passed by reference.`

Either way is fine.

Suggestion:
```php
	 * @param array $block        The block to inject the theme attribute into, and hooked blocks before. Passed by reference.
	 * @param array $parent_block The parent block of the given block. Passed by reference.
```

File: src/wp-includes/blocks.php [hunk 2]
```diff
@@ -858,7 +858,7 @@ function make_after_block_visitor( $hooked_blocks, $context ) {
 	 * @param array $next         The next sibling block of the given block.
 	 * @return string The serialized markup for the given block, with the markup for any hooked blocks appended to it.
 	 */
-	return function ( &$block, $parent_block = null, $next = null ) use ( $hooked_blocks, $context ) {
+	return function ( &$block, &$parent_block = null, $next = null ) use ( $hooked_blocks, $context ) {
```
Comment (hellofromtonya) [inline src/wp-includes/blocks.php:861]: Same here https://github.com/WordPress/wordpress-develop/pull/5598/files#r1377997304.

But also while here, the default value should also be documented using the format of:

```php
Default null.
```

which can be added to the end of each `@param` description, unless too long, then put on the next line.

File: src/wp-includes/blocks.php [hunk 3]
```diff
@@ -1132,6 +1132,7 @@ function traverse_and_serialize_block( $block, $pre_callback = null, $post_callb
  */
 function traverse_and_serialize_blocks( $blocks, $pre_callback = null, $post_callback = null ) {
 	$result = '';
+	$parent = null; // At the top level, there is no parent block to pass to the callbacks; yet the callbacks expect a reference.
```
Comment (hellofromtonya) [inline src/wp-includes/blocks.php]: Is there a reason not to name the same as the callback's parameter name? Keeping the naming same, i.e. `$parent_block`, can help with readability and understanding.

Comment (ockham) [reply src/wp-includes/blocks.php]: No particular reason! Will adjust 👍

File: src/wp-includes/blocks.php (hunk 1 above)
Comment (ockham) [reply src/wp-includes/blocks.php:788]: Addressed in https://github.com/WordPress/wordpress-develop/pull/5598/commits/50a340ab06543379ed062ca526d3f0b4e5d99a6d.

File: src/wp-includes/blocks.php (hunk 2 above)
Comment (ockham) [reply src/wp-includes/blocks.php:861]: Addressed in https://github.com/WordPress/wordpress-develop/commit/50a340ab06543379ed062ca526d3f0b4e5d99a6d and https://github.com/WordPress/wordpress-develop/pull/5598/commits/39c59938ddb50c8e77be7ec14436a1fc7e5eb578.

File: src/wp-includes/blocks.php (hunk 3 above)
Comment (ockham) [reply src/wp-includes/blocks.php]: Addressed in https://github.com/WordPress/wordpress-develop/pull/5598/commits/7aa6c9cd0c99f3281995458fe54effd3d7ccf651.

**Review by hellofromtonya — APPROVED** [review]
As explained, these changes are needed as continued iteration for 6.5 revealed pass by reference is required. Doing these changes avoids a BC break in the 6.5 cycle.

The change also includes a little bit of girl scouting to improve the `@param` documentation.

**Comment by ockham** [comment]
Committed to Core `trunk` in [`[57038]`](https://core.trac.wordpress.org/changeset/57038), and backported to the 6.4 branch in [`[57039]`](https://core.trac.wordpress.org/changeset/57039).
